### ISSUE TITLE
Restoring original wallet leads to loss of funds and claim data

### ISSUE THREAD

**Issue by tzarebczan** [lbryio/lbryum#155]
Out of curiosity, I tested a couple scenarios revolving around wallet backups and their usability in case of wallet loss.  The typical user, cryptocurrency savvy or not, generally backs up their wallet when they initially setup their wallet software, which is, in this case, the LBRY App and associated electrum wallet. Unless otherwise instructed, users would not backup their wallet regularly and would assume that their original backup would be fully functional in the case of wallet loss.  After doing some testing, this does not seem to be the case when it comes to wallet addresses that were generated in the new wallet (outside of the original addresses that are generated upon initial run) and claims (including both channels and published claims).  

The scenario goes as follows:
1) Install and run LBRY on Fresh PC
2) Backup default wallet that is created by LBRY
3) Create a channel
4) Publish some content
5) Generate new wallet addresses (about 10/20)
6) Send LBC to newest generated address
7) Close LBRY 
8) Lose wallet
9) Restore original default_wallet from backup into lbryum folder, delete blockchain_headers
10) Run LBRY

Upon running LBRY with the original wallet backup, you find that
1) Channel is missing
2) Claim are missing in published area, claim_list_mine is blank
3) LBRY sent in step 6 above is missing.

This sounds like a really big issue to me.  We need to figure out what parts of this are recoverable from the original private key in the wallet. If certain parts are not recoverable, we need to prompt users to backup their wallets after certain situations (i.e, after they publish, after they generate X addresses, etc).

**Comment by finer9** [comment]
I think we wait for Jack's pending doc and see how it works and if we can automate it

**Comment by tzarebczan** [comment]
I ran through this a couple of times and can reproduce it successfully.  At first glance, the funds are indeed missing - the wallet will not show any new transactions that happened since the original backup unless those TXs originated from any original funds that were in the backup (but change addresses would not be captured, and thus, the wallet balance is incorrect). If you muck around enough with sending funds/generating addresses, you may get the wallet to recover one of the public keys and some of the balance.  The claims and channels definitely appear to be lost with the original backup. @jackrobison mentioned that the wallet should be able to recover at least the funds portion, but would be curious if there is any way to regenerate the channel key and claims. 

TLDR:
1) Using an original wallet backup leads to what looks like a loss of funds to the user
2) If you don't backup your wallet after creating identities / claims, they are most likely lost.

**Comment by lyoshenka** [comment]
@tzarebczan I think that all you need to do to restore the funds is to recover the wallet and the generate a bunch of new addresses (enough to generate all the addresses that had funds in them). Can you try recovering a wallet and calling getnewaddress a few dozen times? If you have to do anything else to get your credits back, it would be good to know what that is.

**Comment by tzarebczan** [comment]
@lyoshenka I gave this a shot and after generating 100+ addresses, I was unsuccessful at recovering all the funds. Some previous publishes came through as well as incoming funds, but in the end I couldn't get a change address from a previous sent tx to generate/show balance. 

It's pretty much a crapshoot when generating the addresses - I don't think they are done randomly.  

Per @kaykurokawa , electrum should be able to re-subscribe to the addresses somehow, we should look into this when someone has the bandwidth.

**Comment by SuperGeniusZeb** [comment]
I made a backup when I had just 2 LBC, and before I created my channel and uploaded 2 pictures and 2 videos to it. I had to reinstall LBRY, and I when I used my wallet backup, the LBRY app did not show my wallet as having the 24 LBC I had pre-reinstall (instead it just showed 2 LBC), and it did not recognize the ownership of my channel on the "PUBLISH" screen, nor did it show anything under the "PUBLISHED" list. The transaction history showed all (I think) of my "receive" transactions, as well as the "channel" transaction from when I claimed my channel, and also 2 of the "update" transactions for my recent updates to one of my two videos. (I might have made another update to that video, but I'm not sure.) However, not all of the "update" transactions for the first video showed up, none of the "update" transactions for my second video didn't show up, and none of the initial "publish" transactions showed up, either.

I tried clearing the application cache and restarting the app twice, but that didn't fix it.

Up until this point, I don't think I had ever pressed the "GET NEW ADDRESS" button in the LBRY app.

I then decided to use that button, and after pressing it about a dozen times, I managed to get all (I think) of my transactions to show up in the transaction history. I should note that the transactions that were restored by generating new receive addresses showed up at the top of the transaction history, even though the dates listed on the transactions were earlier than the ones below them that were already in the transaction history. Generating a bunch of receive addresses also restored my channel in the channel list in the "PUBLISH" screen, as well as cause the 2 images I had uploaded to show up in the "PUBLISHED" list.

If I had to guess, the videos didn't show up because their "published" transactions were listed after the "update" transactions that were already there, whereas the images had never been updated and didn't show up in the transaction history until I started generating receive addresses.

I then cleared the application cache again, and this fixed the order of transactions in the transaction history, but the 2 videos I had uploaded still didn't show up under "PUBLISHED", and going to the URLs of the videos didn't show a "REMOVE" button like the images now did.

I then restarted the app to see if it would make the videos show up under "PUBLISHED"... it didn't work. However, I pressed the "GET NEW ADDRESS" button a couple more times, and another "update" transaction for one of my 2 videos showed up, and that video then showed up under "PUBLISHED".

I also noticed that there was a bit of delay (the button grayed out for longer than normal) after pressing the "GET NEW ADDRESS" button the second time, and I'm going to take a guess and say that when that happens, that usually means that some transactions were restored that time. This little bit of info might be helpful to anyone trying to spam-click the button to get their transactions to show up again.

I pressed the button a couple more times yet again, and then there was a delay again after pressing it, so I checked, and my other video was now listed under "PUBLISHED", and I had regained control of all 4 of my uploads.

I decided to try clearing the application cache again, and interestingly, this caused my 2 videos to disappear from "PUBLISHED" again. After going to the videos via URL, however, they showed up again. I'm not sure if the videos would have just shown up on their own if I had waited a little longer after clearing the application cache.

None of this restored my missing LBC, however, despite the transaction history showing multiple "receive" transactions. I thought it must be because I was still missing a couple transactions in my transaction history, so I started spam-clicking the "GET NEW ADDRESS" button again.

After several presses, the button grayed out for a long time. A looong time. I wasn't sure if it had frozen up or if it was loading all my lost credits or something, so I decided to leave my computer and come back later and see if anything had happened.

I waited a couple hours, but the button was greyed out still, so I decided to go clear the application cache yet again. I then proceeded to spam click the button again, and after many presses, it greyed out and stayed that way again. I repeated this, and I guess there's a either a bug or some kind of limitation that only lets you generate so many receive addresses before you have to restart the app or clear the application cache.

I should also note that clearing the application cache once again caused my 2 videos to disappear from "PUBLISHED" until I went to their URLs, though I still don't know if they would have showed up on their own if I just waited longer.

Hopefully this info will be helpful to others who have experienced this issue and/or to devs looking to improve how this functions!

**Comment by shyba** [comment]
I could reproduce it and added some tests in this lbrytest branch that might help finding the cause: https://github.com/lbryio/lbrytest/tree/tests_for_backups

**Comment by lyoshenka** [comment]
Any idea how to fix it then? Is it just a matter of generating more
addresses, or is the backup actually missing necessary information?

**Comment by tzarebczan** [comment]
@ZebulanStanphill if you still have this old wallet backup, we have a way to restore it using our new wallet. Please reach out to me on Discord https://chat.lbry.io - @jiggytom

**Comment by tzarebczan** [comment]
This is no longer an issue in the new wallet - it's able to sync from a fresh seed.
